[[ WIP ]] Split 'number' into 'real' and 'integer'.

runrevmark opened the pull request:
The arithmetic defined in this patch works as follows:

There are two concrete numeric types:
- Integer - which can currently hold an integral value in the range [INT32_MIN..UINT32_MAX]
- Real - which can currently hold any IEEE double value

Both of these types are implemented as variants of MCNumberRef.

Integer literals are considered to be of type 'Integer' (this is any sequence of digits without '.' or exponent).
Real literals are considered to be of type 'Real' (real literals must include a '.' or an exponent).

If you perform an operation on two Integers then you will get Integer semantics - in particular, numeric overflow will be detected and an error raised.

If you perform an operation on two Reals then you get Real semantics - in particular, numbers are mostly approximately represented.

If you perform an operation between an Integer and a Real then the Integer is promoted to a Real and you get Real semantics.

Current Real semantics are that if you perform a divide by zero, you will get an error thrown; if an operation results in a number which is not 'finite' (as defined by isfinite()) then you get an error thrown.

Either an Integer or a Real can be assigned to a slot of type 'number', but an Integer cannot be assigned to a slot of type 'real' and a Real cannot be assigned to a slot of type 'integer'.

Syntax needs to be added to allow explicit conversion between real and integer:
- <number> as real: approximate <number> by an IEEE double
- <number> as integer: take the integer part of floor(<number), throwing an overflow error if the resulting value will not fit in Integer's permissable range.

runrevmark committed:
[[ WIP ]] Work on sorting out MCNumberRef.

peter-b commented on libfoundation/src/foundation-number.cpp:
```diff
@@ -207,6 +245,56 @@ bool MCNumberParseUnicodeChars(const unichar_t *p_chars, uindex_t p_char_count,
+            return MCStringFormat(r_string, "%lf", p_number -> real);
+    }
+    
+    MCUnreachable();
```
I think that stylistically it would be better for this to be in the `switch` statement's `default` branch, but YMMV etc. :wink:

runrevmark replied:
Sure - that's probably a better idea.

peter-b commented on libfoundation/src/foundation-number.cpp:
```diff
@@ -207,6 +245,56 @@ bool MCNumberParseUnicodeChars(const unichar_t *p_chars, uindex_t p_char_count,
+    
+    char *t_end;
+    integer_t t_integer;
+    t_integer = strtol((const char *)t_chars, &t_end, 10);
```
Inconveniently, strtol(3) will consume any amount of leading whitespace, but won't consume trailing whitespace.  The `*t_end == '\0'` check means that parsing will barf on trailing whitespace but will ignore leading whitespace.

runrevmark replied:
Hmmm - I wonder if we should not consume whitespace at the start of a string? i.e. The make it strict until there's a reason not too argument...

peter-b replied:
I vote for strictness. :wink:

peter-b commented on libfoundation/src/foundation-number.cpp:
```diff
@@ -216,41 +304,1102 @@ bool __MCNumberCopyDescription(__MCNumber *self, MCStringRef& r_string)
+            return MCHashInteger(self -> integer);
+        case kMCNumberFlagUnsignedIntegerRep:
+            return MCHashInteger((integer_t)self -> unsigned_integer);
+        default:
```
This might be prettier with `double` representation as a specific case in the switch statement, and `MCUnreachable()` in the `default` branch.

peter-b commented on libfoundation/src/foundation-number.cpp:
```diff
@@ -216,41 +304,1102 @@ bool __MCNumberCopyDescription(__MCNumber *self, MCStringRef& r_string)
+{
+    int64_t w;
+    w = (int64_t)x * (int64_t)y;
+    if ((w > INT32_MAX) || (w < INT32_MIN))
```
I checked that `INT32_MIN * INT32_MIN < INT64_MAX` :+1:

peter-b commented on libfoundation/src/foundation-number.cpp:
```diff
@@ -216,41 +304,1102 @@ bool __MCNumberCopyDescription(__MCNumber *self, MCStringRef& r_string)
+{
+    uint64_t w;
+    w = (uint64_t)x * (uint64_t)y;
+    if ((w > INT64_MAX) || (w < INT64_MIN))
```
I think this this test should be:

```
if (w > UINT32_MAX)
```

peter-b commented on libfoundation/src/foundation-number.cpp:
```diff
@@ -216,41 +304,1102 @@ bool __MCNumberCopyDescription(__MCNumber *self, MCStringRef& r_string)
+        return true;
+    }
+    
+    static inline bool unsigned_by_signed(uinteger_t x, integer_t y, integer_t& r_value)
```
There doesn't appear to be the possibility of adding a positive signed integer to a positive unsigned integer resulting in a positive unsigned integer, even though there are a lot of cases where this is an entirely reasonable thing to do and doesn't cause overflow.

peter-b replied:
I suppose that adding "unsigned result" variants of each of the integer ops adds a lot of additional paths to write and test, which isn't ideal.

peter-b replied:
However... it would be nice for this to work (pseudocode):

```
variable tUnsigned as unsigned
variable tNumber
put INTEGER_MAX into tUnsigned
put 2 into tNumber
add tNumber to tUnsigned
```

runrevmark replied:
I'm not sure introducing a special-case 'unsigned' type at the moment would really add much - it gets subsumed by having 'integer range' types. This would help both an optimization phase in the VM to do with using actual integer slots (rather than MCNumberRef's); and with property declarations (as it would be an actual language feature being able to specify the range of a number, rather than metadata as it will be at the moment).

What you've written above works with the current implementation (replace unsigned with integer) - adding 2 to INT_MAX would cause the value of tNumber to go from being a signed rep to an unsigned rep.

peter-b replied:
Okay, I'm happy with that. :+1:

peter-b commented on libfoundation/src/foundation-number.cpp:
```diff
@@ -216,41 +304,1102 @@ bool __MCNumberCopyDescription(__MCNumber *self, MCStringRef& r_string)
+        if (!__checked_unsigned_subtract(y, (uinteger_t)-x, &z))
+            return false;
+        
+        if (z > INT32_MAX)
```
`z` is `integer_t`, so should be compared against `INTEGER_MAX` (etc. for the whole patch).

peter-b commented on libfoundation/src/foundation-number.cpp:
```diff
@@ -216,41 +304,1102 @@ bool __MCNumberCopyDescription(__MCNumber *self, MCStringRef& r_string)
+        // if y <= 0 then
+        //   x + y === x + -(-y) === x - (-y)
+        uinteger_t z;
+        if (!__checked_unsigned_subtract(x, (uinteger_t)-y, &z))
```
If `y` is `INTEGER_MIN` then `-y` will overflow `integer_t` storage **before** the cast to `uinteger_t`.

runrevmark replied:
Oh the subtlety of C :)

Hmmm - I'll abstract that to an inline function which 'does the right thing'.

peter-b commented on libfoundation/src/foundation-number.cpp:
```diff
@@ -216,41 +304,1102 @@ bool __MCNumberCopyDescription(__MCNumber *self, MCStringRef& r_string)
+    }
+};
+
+// The '/' operator always produces reals (use div or mod for integer ops).
```
If the author has specified integer inputs and the output is going into integer storage, it doesn't seem unreasonable to infer that `/` should be interpreted as `div`... :question:

runrevmark replied:
The problem with that is that you don't know what the container type is (polymorphic selection is based on the dynamic type of inputs - inout arguments don't help here since you can be pass in a value as long as one of its supertypes is that target type).

All the other arithmetic operations have the property that the integer forms are 'subsets' (in some sense) of the real forms... Division is different as integer division loses information.

runrevmark replied:
Of course - I just realized you said '/' not 'divide'... 'divide' cannot be implemented in this way for the above reasons, so it seems sensible that '/' should not either.

Also it makes the code 100% clear when you are doing and expecting 'div' and when you are expecting actual division.

peter-b replied:
Other programming languages have survived with a `/` operator that does slightly different things with floats than with integers, but I suppose that it can be argued either way. Let's stick with your approach and see how it works out.

peter-b commented:
Okay, having looked through this and thought about it very hard, I don't think it's useful to be able to store an integer > `INTEGER_MAX` but not to be able to increment it without losing its integer-ness.  I think the answer is to do one of the following:
1. Implement **signed, unsigned → unsigned** and **unsigned, signed → unsigned** ops
2. Remove unsigned type and ops

runrevmark commented:
It shouldn't lose it's integerness - for the number types if the fast path (using integers) fails it goes through the overflow path which produces an integer (unsigned rep if necessary) if it can.

When we add a 64-bit rep then the slow path would happen very rarely...

The other thing to experiment with at some point is splitting on +ve and -ve rather than int and uint that won't fit into an int. Although that would move the point of inflection to around zero rather than Int-max (whatever that might be).
Sent from my iPhone

> On 12 Feb 2015, at 16:59, Peter TB Brett notifications@github.com wrote:
> 
> Okay, having looked through this and thought about it very hard, I don't think it's useful to be able to store an integer > INTEGER_MAX but not to be able to increment it without losing its integer-ness. I think the answer is to do one of the following:
> 
> Implement signed, unsigned → unsigned and unsigned, signed → unsigned ops
> Remove unsigned type and ops
> —
> Reply to this email directly or view it on GitHub.

peter-b commented:
What about storing integers as unsigned integer with external sign bit in the flags word?

peter-b commented:
On second thoughts, that's a silly idea.

I have this overwhelming feeling of wheel-reinvention going on. :-(

runrevmark commented:
Heh - well the 'external sign' bit is a better way to express the split over +ve/-ve. Having thought about this, I don't think a split over zero makes sense as it would make any operation between + and - more expensive (e.g. counting down a large unsigned value with a small negative delta).

As posed the slow path only kicks in on the transition between INTEGER_MAX and INTEGER_MAX + 1 which I wouldn't imagine would be a particularly frequent case. If the range of INTEGER_MAX is taken up to 64-bit levels, then I can only imagine it would be even rarer.

To be fair, the current implementation with templates is far inferior to a hand-crafter per operation implementation which is what would really ensure things were as fast as possible. It also occurs to me that that in this case well-defined 'bit-twiddling' could be used to minimise branching in the mixed cases (unsigned op signed / signed op unsigned). Furthermore, it is only really addition and subtraction which are the key ones here - multiplication and division have very fast exit for signed/unsigned mixes.

Performance-wise, doing assembly-level implementations would be the ideal situation - particularly on 32-bit processors where we'd want to not pay for 64-bit operations until they are needed.

runrevmark committed:
[[ WIP ]] Made use of switch / default / unreachable more consistent and ensured that all possible valid enum values are used in cases.

runrevmark committed:
[[ WIP ]] Change uses of UINT32/INT32_MIN/MAX to UINTEGER/INTEGER_MIN/MAX.

peter-b commented:
I can't think of any C compiler that wouldn't emit highly efficient code for these types of twiddling. It probably wouldn't be worth our time to write a set of assembly-level ops.

runrevmark committed:
[[ WIP ]] Make TryToParse more strict and correctly handle non-native StringRefs.

runrevmark committed:
[[ WIP ]] Correct validity check in __checked_unsigned_multiply.

runrevmark committed:
[[ WIP ]] Abstract negative signed integer negation into __negate_negative_signed (not sure the 'correct' way to express the operation).

runrevmark committed:
[[ WIP ]] Added MCNumberIsValid API and cleaned up internal rep type checking.

runrevmark committed:
[[ WIP ]] Abstracted errors (numeric overflow and division by zero).

runrevmark committed:
[[ WIP ]] Add MCNumberIsFinite() check along with MCNumberFinite*() operations.
[[ WIP ]] Updated LCB arithmetic module to use Finite operations, and block infinite values from strings.

runrevmark committed:
[[ WIP ]] Fix failure to check for nil pointer in return from MCNumberTryToParse().
[[ WIP ]] Correct test.mlc.

peter-b commented:
Looking pretty solid to me! :+1:

peter-b commented on libfoundation/src/foundation-number.cpp:
```diff
@@ -216,41 +382,1247 @@ bool __MCNumberCopyDescription(__MCNumber *self, MCStringRef& r_string)
+        if (!__checked_unsigned_subtract(y, x, &z))
+            return false;
+        
+        if (z > -INTEGER_MIN)
```
This unfortunately overflows. I think you need:

```
if (z > -(int64_t)INTEGER_MIN)
```

runrevmark committed:
[[ WIP ]] Fix overflow warnings when doing -INTEGER_MIN.
[[ WIP ]] Correct __negate_negative_signed() op.

runrevmark committed:
Merge branch 'develop' into feature-lcb_proper_integers

runrevmark committed:
[[ WIP ]] Better name for __MCNumberBinaryType - now __MCNumberBinaryOpType.

runrevmark committed:
[[ WIP ]] Fix codegen bug in repeat up to and ensure builtin signature of repeatupto/downto is correct.

runrevmark committed:
[[ WIP ]] Implemented real-only variants of number operations.
[[ WIP ]] Changed things so that integer < number and real < number but integer is not related to real.

The arithmetic in LCB now works as follows:

There are two concrete numeric types:
  Integer - which can currently hold an integral value in the range [INT32_MIN..UINT32_MAX]
  Real - which can currently hold any IEEE double value

Both of these types are implemented as variants of MCNumberRef.

Integer literals are considered to be of type 'Integer' (this is any sequence of digits without '.' or exponent).
Real literals are considered to be of type 'Real' (real literals must include a '.' or an exponent).

If you perform an operation on two Integers then you will get Integer semantics - in particular, numeric overflow will be detected and an error raised.

If you perform an operation on two Reals then you get Real semantics - in particular, numbers are mostly approximately represented.

If you perform an operation between an Integer and a Real then the Integer is promoted to a Real and you get Real semantics.

Either an Integer or a Real can be assigned to a slot of type 'number', but an Integer cannot be assigned to a slot of type 'real' and a Real cannot be assigned to a slot of type 'integer'.

Syntax needs to be added to allow explicit conversion between real and integer:
  - <number> as real: approximate <number> by an IEEE double
  - <number> as integer: take the integer part of floor(<number), throwing an overflow error if the resulting value will not fit in Integer's permissable range.

runrevmark commented:
It occurs to me that the current model outlined above only really has two surprising elements...

The first is the strict distinction between 1 and 1.0 - this is only noticeable if you use integer or real typed vars (not number) as you'll get a typecheck error if you pass the wrong type of constant to the opposite type variable - but then, you'd get a warning in other languages for the same thing.

The second is that (if using number types) passing just integers through will cause an overflow error potentially; whereas reals will silently approximate. This problem goes away if integers can be multiple precision. However, that being said, the semantics of integer and real are really quite different so perhaps it is the number type idea which is slightly flawed - maybe you should have to explicitly choose between the two rather than allowing a general 'number' type.

livecodeian commented on libfoundation/src/foundation-number.cpp:
```diff
@@ -207,6 +306,73 @@ bool MCNumberParseUnicodeChars(const unichar_t *p_chars, uindex_t p_char_count,
+    return false;
+}
+
+bool MCNumberTryToParse(MCStringRef p_string, MCNumberRef& r_number)
```
Can you add a variant of this function that accepts an offset to begin parsing from, and that returns the number of characters consumed rather than throwing an error if the whole string isn't used?

runrevmark replied:
How about:

```
bool MCNumberTryToParseRange(MCStringRef p_string, MCRange p_range, MCRange *r_optional_out_range, MCNumberRef& r_number)
```

This attempts to parse 'p_range' of p_string as a number. If 'p_out_range' is specified then it will return the range of characters that were parsed as the number without an end-boundary check. If 'p_out_range' is nil, then a end-boundary check is done.

livecodeian replied:
OK, I can use that :)

runrevmark committed:
[[ WIP ]] Restructure arithmetic ops.

runrevmark committed:
[[ WIP ]] More work on restructuring operations.

peter-b commented:
I see that a `LIKELY()` macro has appeared -- I suspect this should be called `MC_LIKELY()` and defined in `foundation.h`...

runrevmark committed:
[[ WIP ]] Finished restructuring.

Integer rep can now be 32-bit or 64-bit.

As much common code has been factored out - experimental use of 'force inline' to make sure maximum optimization potential.

runrevmark committed:
[[ WIP ]] Further work.

Updated header.

Added use of CLANG builtins for overflow detecting ops.

runrevmark committed:
Merge branch 'develop' into feature-lcb_proper_integers

runrevmark committed:
[[ WIP ]] Further work.

Added generation utility for test data file.

Added tests for integer + - * mod div.

Reinstated 'wrap' syntax.

runrevmark committed:
[[ WIP ]] Arithmetic integer tables.

Temporary commit of 32-bit internal rep integer testing tables.

runrevmark committed:
Merge branch 'develop' into feature-lcb_proper_integers

runrevmark committed:
Merge branch 'bugfix-14695' into feature-lcb_proper_integers

runrevmark committed:
[[ Foundation ]] MCHandlerTryToInvokeWithList API

New API call which invokes a handler value with a list of arguments. If the invoke throws an error, it is returned in the result rather than propagating it.

runrevmark committed:
[[ Foundation ]] MCNumberTryToParse does not fail on empty string.

runrevmark committed:
[[ Foundation ]] MCNumber errors not named.

The MCNumber error typeinfo's are now correctly bound to a named typeinfo.

runrevmark committed:
[[ Foundation ]] MCErrorCreateWithMessage not using message.

The MCErrorCreateWithMessage API now correctly propagates the message string to the constructed error object.

runrevmark committed:
[[ WIP ]] Added testing of throwing integer operations.

runrevmark committed:
[[ WIP ]] Remove extraneous test diagnostics.

livecode-vulcan commented:
📝 Hi @runrevmark, I haven't been able to verify that you've signed our [LiveCode Contributor's Agreement](https://livecode.com/account/developer/contribute).

If you have previously signed the Contributor's Agreement, I may not be able to detect it because you haven't [linked your GitHub account to your LiveCode account](https://livecode.com/account/developer/github).

Please see the [information for contributors](https://github.com/runrev/livecode/blob/develop/HACKING.md) for more information.
